core/iwasm/common/wasm_runtime_common.c

```diff
@@ -2310,10 +2311,8 @@ wasm_set_exception(WASMModuleInstance *module_inst, const char *exception)
     WASMExecEnv *exec_env = NULL;
 
 #if WASM_ENABLE_SHARED_MEMORY != 0
-    WASMSharedMemNode *node =
-        wasm_module_get_shared_memory((WASMModuleCommon *)module_inst->module);
-    if (node)
-        os_mutex_lock(&node->shared_mem_lock);
+    if (module_inst->memory_count > 0)
+        shared_memory_lock(module_inst->memories[0]);
```

Comment:
@wenyongh Do you remember why we need to lock the shared memory when setting exception?

Reply:
see https://github.com/bytecodealliance/wasm-micro-runtime/issues/2407

Reply:
yes, I remember that one instance's exception may be accessed by other threads when spreading non "wasi proc exit" exception. It was introduced in this PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/1995

In thread_manager.c, function set_exception_visitor:
```C
        /* Only spread non "wasi proc exit" exception */
#if WASM_ENABLE_SHARED_MEMORY != 0
        WASMSharedMemNode *shared_mem_node = wasm_module_get_shared_memory(
            (WASMModuleCommon *)curr_wasm_inst->module);
        if (shared_mem_node)
            os_mutex_lock(&shared_mem_node->shared_mem_lock);
#endif
        if (!strstr(wasm_inst->cur_exception, "wasi proc exit")) {
            bh_memcpy_s(curr_wasm_inst->cur_exception,
                        sizeof(curr_wasm_inst->cur_exception),
                        wasm_inst->cur_exception,
                        sizeof(wasm_inst->cur_exception));
        }
#if WASM_ENABLE_SHARED_MEMORY != 0
        if (shared_mem_node)
            os_mutex_unlock(&shared_mem_node->shared_mem_lock);
#endif
```

Reply:
> yes, I remember that one instance's exception may be accessed by other threads when spreading non "wasi proc exit" exception. It was introduced in this PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/1995

Yes, that's the reason, I was about to comment. When a thread runs into an exception it spreads it to other threads by using `set_exception_visitor` and that other threads could try to read it at the same time, causing concurrency issues.

Reply:
Thanks a lot!